server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1211,6 +1241,263 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
         );
     }
 
+    /**
+     * Maximum number of on-going snapshots to delete per repository
+     */
+    public static final Setting<Integer> MAX_CONCURRENT_SNAPSHOT_TO_DELETE_PER_REPOSITORY_SETTING = Setting.intSetting(
+        "snapshot.snapshots_to_delete.max_concurrent_operations",
+        5,
+        1,
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
+    );
+
+    public static final Setting<TimeValue> SNAPSHOT_TO_DELETE_RETRY_INTERVAL_SETTING = Setting.timeSetting(
+        "snapshot.snapshots_to_delete.retry_interval",
+        TimeValue.timeValueSeconds(30L),
+        TimeValue.timeValueSeconds(1L),
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
+    );
+
+    /**
+     * Mutex used to modify snapshots to delete related objects.
+     */
+    private final Object snapshotToDeleteMutex = new Object();
+
+    /**
+     * Set of snapshots to delete whose deletion will be triggered, per repository
+     */
+    private final Map<String, Set<SnapshotId>> onGoingSnapshotsDeletions = new HashMap<>();
+
+    /**
+     * Queue of snapshots whose deletion must be triggered. Each repository has a dedicated queue.
+     */
+    private final Map<String, Queue<SnapshotId>> queuesSnapshotsDeletions = new HashMap<>();
+
+    /**
+     * Number of on-going snapshots to delete per repository
+     */
+    private final Map<String, Integer> numberOfOnGoingSnapshotsToDeletes = new HashMap<>();
+
+    private volatile int maxConcurrentSnapshotsToDeletes;
+    private volatile TimeValue snapshotsToDeleteRetryInterval;
```

Comment:
Detecting conflicting situations and not trigger snapshot deletions sounds like a good idea, thanks. I refactored the deletion logic in 3e8dc653f9d939ba35857dc704f66a036ee5c990 to detect conflicts and do nothing but rely on subsequent cluster state updates to allow and to trigger snapshot deletions.

Still, I went back and forth with the time base retry. I agree this should not be the main path but I think that it will be useful in some situations, like the snapshot deletions failing because of expired repository or network issue when accessing the repository. I made an infinite retry with 30s delay but we could imagine some backoff policy or just a max number of retries before giving up.